### PR TITLE
storage-client: track since frontiers for rehydration

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

The rehydrating client was previously ignoring `AllowCompaction` commands which meant that new clusters were only receiving `CreateSource` commands for the sources that the rehydrating client was tracking.

For the most part this worked correctly, because the first time the `AllowCompaction` command was being sent by the controller, it would reach the cluster which would cleanup the dataflow, and then would respond with a `DroppedIds` response that would lead to the rehydrating client forgetting about a source, and therefore not re-sending it to new clusters in future rehydrations.

In the not-happy case though, an `AllowCompaction` command could be sent by the storage controller but the cluster could crash before it could send the `DroppedIds` response. In this situation the rehydrating client would forever believe that the source for which the `AllowCompaction` command was lost should be running. The reason for this is that the controller had already send the original `AllowCompaction` command, so from its perspective there is nothing else to do, and the new cluster that got rehydrated only heard about a `CreateSource` command so from its perspective the source has not been dropped.

The solution implemented in this PR is for the rehydrating client to not throw away the `AllowCompaction` commands but instead to keep them in its state, just like we do for every other command, and re-send them any time we want to rehydrate a client. This guarantees that no matter how many crashes happen a cluster will eventually process all the pending commands and will correctly generate a `DroppedIds` response, to clean up the state of all the clients.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-client/src/controller/rehydration.rs [hunk 1]
```diff
@@ -139,6 +140,8 @@ struct RehydrationTask<T> {
     sinks: BTreeMap<GlobalId, CreateSinkCommand<T>>,
     /// The upper frontier information received.
     uppers: BTreeMap<GlobalId, Antichain<T>>,
+    /// The since frontiers that have been observed.
```
Comment (aljoscha) [inline src/storage-client/src/controller/rehydration.rs:143]: Might be worthwhile to highlight that uppers come back from the cluster, while sinces "go out" from the controller.

**Review by aljoscha — CHANGES_REQUESTED** [review]
I still have a slight suspicion that we could fail to report back dropped IDs in almost all cases because of how this is structured: https://github.com/MaterializeInc/materialize/blob/917ce78e739e6f8b8b41e602f33e6a8a6646d805/src/storage/src/storage_state.rs#L435

After we receive the `CreateSources` and the `AllowCompaction`, the drop responses are stashed in `dropped_ids`. And then we do one timely step, which would run the just-created source which would then panic.

This would probably be fixed by moving the block that reports back dropped IDs to after where we call `handle_internal_storage_command`? I hope!

**Comment by guswynn** [comment]
@aljoscha We would still need to handle the case where we drop a source, then IMMEDIATELY drop a cluster right after it. the two `DROP` statements can be planned and executed before the `DroppedId`'s response is crafted, which can no longer be sent because the client connection is broken

**Comment by petrosagg** [comment]
superseded by https://github.com/MaterializeInc/materialize/pull/18115

**Pull request closed without merging**